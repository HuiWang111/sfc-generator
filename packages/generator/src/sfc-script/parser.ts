import type { ParserOptions } from '@babel/parser'
import type { ScriptParseOptions } from '../types'
import { parse as babelParse } from '@babel/parser'
import traverse from '@babel/traverse'
import { CompositionApi } from './composition-api'
import { OptionsApi } from './options-api'

export function parse<T extends boolean = false>(
  code: string,
  options: ScriptParseOptions<T> = {},
): {
    api: T extends true ? CompositionApi : OptionsApi
    ast: ReturnType<typeof babelParse>
  } {
  const { lang, jsx = false, setup = false } = options

  const plugins: ParserOptions['plugins'] = []
  if (lang === 'ts') {
    plugins.push('typescript')
  }
  if (jsx) {
    plugins.push('jsx')
  }

  const ast = babelParse(code, {
    sourceType: 'module',
    plugins,
  })

  let api: OptionsApi | CompositionApi | null = null

  if (setup) {
    api = new CompositionApi(ast.program.body)
  }
  else {
    traverse(ast, {
      ExportDefaultDeclaration(path) {
        if (setup)
          return

        const declarationPath = path.get('declaration')
        if (declarationPath.isObjectExpression()) {
          api = new OptionsApi(declarationPath)
        }
      },
    })
  }

  if (!api) {
    throw new Error(`script setup attr is ${setup}, parse code error: ${code}`)
  }

  return {
    api: api as any,
    ast,
  }
}
