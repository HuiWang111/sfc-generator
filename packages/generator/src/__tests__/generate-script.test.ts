import { parse } from '@babel/parser'
import { generateScript } from '../index'

describe('test generateScript', () => {
  it('generateScript should work', () => {
    const code = `import { ref } from 'vue'

const visible = ref(false)
function setVisible (v) {
  visible.value = v
}`

    const ast = parse(code, {
      sourceType: 'module',
    })
    const res = generateScript(ast)
    expect(res).toMatchSnapshot()
  })

  it('generateScript should work with script attrs', () => {
    const code = `import { ref } from 'vue'

const visible = ref(false)
function setVisible (v) {
  visible.value = v
}`

    const ast = parse(code, {
      sourceType: 'module',
    })
    const res = generateScript(ast, { lang: 'ts', setup: true })
    expect(res).toMatchSnapshot()
  })
})
