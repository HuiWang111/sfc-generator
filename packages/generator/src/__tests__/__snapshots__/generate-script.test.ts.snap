// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test generateScript generateScript should work 1`] = `
"<script>
import { ref } from 'vue';
const visible = ref(false);
function setVisible(v) {
  visible.value = v;
}
</script>"
`;

exports[`test generateScript generateScript should work with script attrs 1`] = `
"<script lang="ts" setup>
import { ref } from 'vue';
const visible = ref(false);
function setVisible(v) {
  visible.value = v;
}
</script>"
`;
